ext/mbstring/mbstring.c

```diff
@@ -3183,77 +3183,368 @@ static int *make_conversion_map(HashTable *target_hash, int *convmap_size)
+
+	mb_convert_buf buf;
+	mb_convert_buf_init(&buf, in_len, MBSTRG(current_filter_illegal_substchar), MBSTRG(current_filter_illegal_mode));
+
+	/* Decode input string from bytes to wchars one 128-wchar buffer at a time, then deconvert HTML entities,
+	 * copying the deconverted wchars to a second buffer, then convert back to original encoding from the
+	 * 2nd 'converted' buffer.
+	 *
+	 * Tricky part: an HTML entity might be truncated at the end of the wchar buffer; the remaining
+	 * part could come in the next buffer of wchars. To deal with this problem, when we find what looks
+	 * like an HTML entity, we scan to see if it terminates before the end of the wchar buffer or not.
+	 * If not, we copy it to the beginning of the wchar buffer, and tell the input conversion routine
+	 * to store the next batch of wchars after it.
+	 *
+	 * Optimization: Scan for &, and if we don't find it anywhere, don't even bother copying the
+	 * wchars from the 1st buffer to the 2nd one.
+	 *
+	 * 'converted_buf' is big enough that the deconverted wchars will *always* fit in it, so we don't
+	 * have to do bounds checks when writing wchars into it.
+	 */
+
+	unsigned int wchar_buf_offset = 0;
+
+	while (in_len) {
+		/* Leave space for sentinel at the end of the buffer */
+		size_t out_len = encoding->to_wchar(&in, &in_len, wchar_buf + wchar_buf_offset, 127 - wchar_buf_offset, &state);
+		out_len += wchar_buf_offset;
+		ZEND_ASSERT(out_len <= 127);
+		wchar_buf[out_len] = '&'; /* Sentinel */
+
+		uint32_t *p, *converted;
+
+		/* Scan for & first
+		 * If wchar_buf_offset > 0, then definitely & will be there, so don't bother in that case */
+		if (wchar_buf_offset == 0) {
+			p = wchar_buf;
+			while (*p != '&')
+				p++;
+			if (p == wchar_buf + out_len) {
+				/* No HTML entities in this buffer */
+				encoding->from_wchar(wchar_buf, out_len, &buf, !in_len);
+				continue;
+			} else {
+				/* Copy over the prefix with no & which we already scanned */
+				memcpy(converted_buf, wchar_buf, (p - wchar_buf) * 4);
+				converted = converted_buf + (p - wchar_buf);
+			}
+		} else {
+			p = wchar_buf;
+			converted = converted_buf;
+		}
+
+found_ampersand:
+		ZEND_ASSERT(*p == '&');
+		uint32_t *p2 = p;
+
+		if (*++p2 == '#') {
+			if (*++p2 == 'x') {
+				/* Possible hex entity */
+				uint32_t w = *++p2;
+				while ((w >= '0' && w <= '9') || (w >= 'A' && w <= 'F') || (w >= 'a' && w <= 'f'))
+					w = *++p2;
+				if (p2 == wchar_buf + out_len) {
+					/* We hit the end of the buffer while still reading digits */
+					if ((p2 - p) <= 11 && in_len) {
+						/* The number of digits was legal (1-8 hex digits)
+						 * We need to process this identity again on the next iteration of the main loop */
+						memmove(wchar_buf, p, (p2 - p) * 4);
+						wchar_buf_offset = p2 - p;
+						goto process_converted_wchars;
+					} else {
+						/* Invalid entity */
+						memcpy(converted, p, (p2 - p) * 4);
+						converted += p2 - p;
+						goto process_converted_wchars_no_offset;
```

Comment:
Well, if we hit the end of buffer, we don't know how many more digits may be at the beginning of the next buffer, so we don't necessarily know if the entity is valid or not. If it *might* be valid, we must reprocess the entity after refilling the wchar buffer.

What if we know that it *definitely is not* valid? Could we still reprocess it on the next iteration? That would work, as long as there are still more unprocessed bytes in the input, since the test for continuing the main loop is `while (in_len)`.

What if we amend that test to `while (in_len || wchar_buf_offset)`? Then we could get stuck in an infinite loop if the entity is at the end of the overall input.

I understand the desire to combine some of the `if` clauses, because there are many of them. I don't see how to do that without breaking one case or another, though.